File: Modules/AcademicPerformance/Service/Api/V1/Contracts/AcademicPerformanceContracts.cs
using Serenity.Services;

namespace AcademicCollectorDemo.Modules.AcademicPerformance.Api.V1.Contracts;

public sealed class AcademicDataCollectRequest : ServiceRequest
{
    public string? Orcid { get; set; } = null;
    public string? GoogleScholarId { get; set; } = null;
    public string? WebOfScienceResearcherId { get; set; } = null;
}

public sealed class AcademicResearcherRequest : ServiceRequest
{
    public int? ResearcherId { get; set; } = null;
    public string? Orcid { get; set; } = null;
    public string? GoogleScholarId { get; set; } = null;
    public string? WebOfScienceResearcherId { get; set; } = null;
}

public sealed class AcademicPublicationListRequest : ServiceRequest
{
    public int? ResearcherId { get; set; } = null;
    public string? Orcid { get; set; } = null;
    public string? GoogleScholarId { get; set; } = null;
    public string? WebOfScienceResearcherId { get; set; } = null;
    public bool ApprovedOnly { get; set; }
    public string? SearchText { get; set; } = null;
    public int Skip { get; set; }
    public int Take { get; set; } = 100;
}

public sealed class AcademicDataResponse : ServiceResponse
{
    public AcademicResearcherDto? Researcher { get; set; } = null;
    public bool IsSaved { get; set; }
    public int PublicationCount { get; set; }
    public string? DatabaseProvider { get; set; } = null;
    public DateTime CollectedAt { get; set; }
    public List<string> Messages { get; set; } = [];
}

public sealed class AcademicResearcherDto
{
    public int Id { get; set; }
    public string? FirstName { get; set; } = null;
    public string? LastName { get; set; } = null;
    public string? AcademicTitle { get; set; } = null;
    public string? Department { get; set; } = null;
    public string? Orcid { get; set; } = null;
    public string? GoogleScholarId { get; set; } = null;
    public string? WebOfScienceResearcherId { get; set; } = null;
    public string? YoksisResearcherId { get; set; } = null;
    public DateTime? LastUpdatedAt { get; set; } = null;
    public OrcidProfileSummaryDto? OrcidProfile { get; set; } = null;
    public GoogleScholarProfileSummaryDto? GoogleScholarProfile { get; set; } = null;
    public OpenAlexProfileSummaryDto? OpenAlexProfile { get; set; } = null;
    public WebOfScienceProfileSummaryDto? WebOfScienceProfile { get; set; } = null;
}

public sealed class OpenAlexProfileSummaryDto
{
    public string OpenAlexAuthorId { get; set; } = string.Empty;
    public string? DisplayName { get; set; } = null;
    public string? LastKnownInstitution { get; set; } = null;
    public int WorksCount { get; set; }
    public int CollectedWorksCount { get; set; }
    public int CitedByCount { get; set; }
    public int? HIndex { get; set; } = null;
    public int? I10Index { get; set; } = null;
    public decimal? TwoYearMeanCitedness { get; set; } = null;
    public DateTime LastUpdatedAt { get; set; }
}

public sealed class GoogleScholarProfileSummaryDto
{
    public string? DisplayName { get; set; } = null;
    public string? Affiliations { get; set; } = null;
    public string? University { get; set; } = null;
    public string? ProfileUrl { get; set; } = null;
    public int? CitationCount { get; set; } = null;
    public int? CitationCountRecent { get; set; } = null;
    public int? HIndex { get; set; } = null;
    public int? HIndexRecent { get; set; } = null;
    public int? I10Index { get; set; } = null;
    public int? I10IndexRecent { get; set; } = null;
    public int? MetricsSinceYear { get; set; } = null;
    public int DocumentsCount { get; set; }
    public DateTime LastUpdatedAt { get; set; }
}

public sealed class OrcidProfileSummaryDto
{
    public string? DisplayName { get; set; } = null;
    public string? CurrentOrganization { get; set; } = null;
    public string? CurrentDepartment { get; set; } = null;
    public string? CurrentRoleTitle { get; set; } = null;
    public int WorksCount { get; set; }
    public int EmploymentsCount { get; set; }
    public int EducationsCount { get; set; }
    public int FundingsCount { get; set; }
    public int PeerReviewsCount { get; set; }
    public DateTime? RecordLastModifiedAt { get; set; } = null;
    public DateTime LastUpdatedAt { get; set; }
}

public sealed class WebOfScienceProfileSummaryDto
{
    public string? DisplayName { get; set; } = null;
    public string? PrimaryOrganization { get; set; } = null;
    public int? HIndex { get; set; } = null;
    public int DocumentsCount { get; set; }
    public int? TotalTimesCited { get; set; } = null;
    public int? TotalCitingPublications { get; set; } = null;
    public int PeerReviewsCount { get; set; }
    public DateTime LastUpdatedAt { get; set; }
}

public sealed class AcademicPublicationListResponse : ServiceResponse
{
    public int ResearcherId { get; set; }
    public List<AcademicPublicationDto> Entities { get; set; } = [];
    public int TotalCount { get; set; }
    public int Skip { get; set; }
    public int Take { get; set; }
}

public sealed class AcademicPublicationSelectionRequest : ServiceRequest
{
    public int ResearcherId { get; set; }
    public List<int> PublicationIds { get; set; } = [];
}

public sealed class AcademicPublicationSelectionResponse : ServiceResponse
{
    public int ResearcherId { get; set; }
    public List<int> PublicationIds { get; set; } = [];
    public int ApprovedCount { get; set; }
}

public sealed class AcademicPublicationDto
{
    public int Id { get; set; }
    public string Title { get; set; } = string.Empty;
    public int? PublicationYear { get; set; } = null;
    public string? Doi { get; set; } = null;
    public string Category { get; set; } = string.Empty;
    public string? Authors { get; set; } = null;
    public string? Publication { get; set; } = null;
    public string? PublicationUrl { get; set; } = null;
    public string Sources { get; set; } = string.Empty;
    public bool IsApprovedForDisplay { get; set; }
}

File: Modules/AcademicPerformance/WebClient/Contracts/AcademicPerformanceContracts.ts
import type { ServiceResponse } from "@serenity-is/corelib";

export interface PublicationSummaryRow {
    Id?: number;
    PersonelID?: string;
    Title?: string;
    PublicationYear?: number;
    Category?: string;
    Authors?: string;
    Publication?: string;
    Doi?: string;
    PublicationUrl?: string;
    Sources?: string;
    IsApprovedForDisplay?: boolean;
}

export interface PublicationDisplayApprovalResponse extends ServiceResponse {
    PersonelID?: string;
    PublicationSummaryIds?: number[];
    ApprovedCount?: number;
}

export interface ResearcherCollectResponse extends ServiceResponse {
    Researcher?: {
        PersonelID?: string;
        FirstName?: string;
        LastName?: string;
        OrcidProfile?: {
            DisplayName?: string;
            CurrentOrganization?: string;
            WorksCount?: number;
            EmploymentsCount?: number;
            EducationsCount?: number;
            FundingsCount?: number;
            PeerReviewsCount?: number;
            RecordLastModifiedAt?: string;
            LastUpdatedAt?: string;
        };
        GoogleScholarProfile?: {
            DisplayName?: string;
            Affiliations?: string;
            University?: string;
            ProfileUrl?: string;
            CitationCount?: number | null;
            CitationCountRecent?: number | null;
            HIndex?: number | null;
            HIndexRecent?: number | null;
            I10Index?: number | null;
            I10IndexRecent?: number | null;
            MetricsSinceYear?: number | null;
            DocumentsCount?: number | null;
            LastUpdatedAt?: string;
        };
        OpenAlexProfile?: {
            OpenAlexAuthorId?: string;
            DisplayName?: string;
            LastKnownInstitution?: string;
            WorksCount?: number;
            CollectedWorksCount?: number;
            CitedByCount?: number;
            HIndex?: number;
            I10Index?: number;
            TwoYearMeanCitedness?: number;
            LastUpdatedAt?: string;
        };
        ScopusProfile?: {
            ScopusAuthorId?: string;
            DisplayName?: string;
            CurrentAffiliation?: string;
            DocumentsCount?: number;
            CollectedWorksCount?: number;
            CitationCount?: number;
            CitedByCount?: number;
            HIndex?: number;
            LastUpdatedAt?: string;
        };
        TrDizinProfile?: {
            Orcid?: string;
            AuthorId?: number;
            DisplayName?: string;
            PublicationCount?: number | null;
            CitationCount?: number | null;
            ProjectCandidateCount?: number;
            ProjectMatchedCount?: number;
            ProjectUnmatchedCount?: number;
            ProjectSearchComplete?: boolean;
            Projects?: TrDizinProject[];
            LastUpdatedAt?: string;
        };
        WebOfScienceProfile?: {
            DisplayName?: string;
            PrimaryOrganization?: string;
            HIndex?: number;
            DocumentsCount?: number;
            WosDocumentsCount?: number;
            WokDocumentsCount?: number;
            TotalTimesCited?: number;
            TotalCitingPublications?: number;
            PeerReviewsCount?: number;
            LastUpdatedAt?: string;
        };
        ProviderDetails?: ResearcherProviderDetails;
    };
    IsSaved?: boolean;
    FailureCode?: string;
    PublicationCount?: number;
    YoksisPublicationCount?: number;
    CategoryMetrics?: AcademicCategoryMetric[];
    Activities?: AcademicActivity[];
    Messages?: string[];
    ProviderFeedback?: ProviderCollectionFeedback[];
}

export interface AcademicActivity {
    Category?: string;
    Provider?: string;
    Title?: string | null;
    Date?: string | null;
    Organization?: string | null;
    Role?: string | null;
    SourceId?: string | null;
}

export interface ResearcherProviderDetails {
    Orcid?: {
        Activities?: unknown;
        OtherNames?: unknown;
        Emails?: unknown;
    };
    OpenAlex?: {
        ExternalIdentifiers?: unknown;
        AlternativeNames?: unknown;
        RawAuthorNames?: unknown;
        Affiliations?: unknown;
        LastKnownInstitutions?: unknown;
        Topics?: unknown;
        TopicShare?: unknown;
    };
    Scopus?: {
        Orcid?: string;
        NameVariants?: unknown;
        CurrentAffiliation?: unknown;
        AffiliationHistory?: unknown;
        SubjectAreas?: unknown;
        CoauthorCount?: number;
    };
    GoogleScholar?: {
        Interests?: unknown;
        CoAuthors?: unknown;
        PublicAccess?: unknown;
        Thumbnail?: string;
    };
}

export interface TrDizinProject {
    Id?: string;
    ProjectNumber?: string | null;
    Title?: string | null;
    StartedDate?: string | null;
    EndDate?: string | null;
    ProjectGroup?: string | null;
    ResearchersJson?: string | null;
    Duty?: string | null;
    AbstractsJson?: string | null;
    KeywordsJson?: string | null;
    OutputsJson?: string | null;
    AttachmentsJson?: string | null;
}

export interface ProviderCollectionReason {
    Code?: string;
    Description?: string;
    AffectedCount?: number | null;
}

export interface ProviderCollectionFeedback {
    Provider?: string;
    Status?: string;
    Unit?: string;
    RetrievedCount?: number;
    RetainedCount?: number | null;
    ExpectedCount?: number | null;
    Reasons?: ProviderCollectionReason[];
}

export interface ResearcherMetricsResponse extends ServiceResponse {
    PersonelID?: string;
    RecalculatedAt?: string;
}

export interface YoksisOperationResult {
    CategoryName?: string;
    OperationName?: string;
    IsSuccess?: boolean;
    RecordCount?: number;
    ExpectedDetailCount?: number;
    RetrievedDetailCount?: number;
    FailedDetailCount?: number;
    FailureReasons?: YoksisFailureSummary[];
    Errors?: string[];
}

export interface AcademicCategoryMetric {
    CategoryName?: string;
    OperationName?: string;
    RecordCount?: number;
}

export interface YoksisFailureSummary {
    Code?: string;
    Description?: string;
    AffectedCount?: number;
}

export interface YoksisCollectResponse extends ServiceResponse {
    PersonelID?: string;
    ResearcherDisplayName?: string;
    IsSaved?: boolean;
    IsDisabled?: boolean;
    IsCached?: boolean;
    YoksisRecordCount?: number;
    YoksisPublicationCount?: number;
    PublicationSummaryCount?: number;
    SuccessfulCategoryCount?: number;
    FailedCategoryCount?: number;
    TotalRecordCount?: number;
    PublicationDetailTotalCount?: number | null;
    PublicationDetailRetrievedCount?: number;
    PublicationDetailFailedCount?: number;
    PublicationFailureReasons?: YoksisFailureSummary[];
    FailureReasons?: YoksisFailureSummary[];
    StopReason?: string;
    Messages?: string[];
    Categories?: YoksisOperationResult[];
}
